src/hooks/progress/use-update-tutorial-progress.ts

```diff
@@ -0,0 +1,143 @@
+import { useEffect, useState } from 'react'
+import { useInView } from 'react-intersection-observer'
+import useAuthentication from 'hooks/use-authentication'
+// Tutorial progress utilities
+import { progressLabelToPercent } from 'lib/learn-client/api/progress/formatting'
+import {
+	useTutorialProgress,
+	useTutorialProgressMutations,
+	TutorialProgressMutationArgs,
+} from 'hooks/progress'
+// Types
+import {
+	TutorialIdCollectionId,
+	TutorialProgressLabel,
+} from 'lib/learn-client/types'
+
+type IntersectionRef = (node?: Element) => void
+
+interface UpdateProgressReturnInterface {
+	startRef: IntersectionRef
+	endRef: IntersectionRef
+}
+
+/**
+ * Given a tutorialId and collectionId,
+ * Return two refs in an object, { startRef, endRef },
+ * which can be placed on elements to track tutorial progress.
+ *
+ * On initial load, regardless of startRef or endRef visibility,
+ * if the tutorial does not yet have recorded progress,
+ * the tutorial will be marked as `visited` (0 percent).
+ *
+ * When the startRef element moves out of the viewport,
+ * the tutorial will be marked as `in_progress` (50 percent).
+ *
+ * When the endRef element moves into the viewport,
+ * the tutorial will be marked as `complete` (100 percent).
+ */
+export function useUpdateTutorialProgress({
+	tutorialId,
+	collectionId,
+}: TutorialIdCollectionId): UpdateProgressReturnInterface {
+	// We don't try to update progress unless we're authenticated
+	const { isAuthenticated } = useAuthentication()
+	// We need to know if progress exists, to know whether to "create" or "update"
+	const { tutorialProgressLabel } = useTutorialProgress({
+		tutorialId,
+		collectionId,
+	})
+	// `react-query` mutations are what we'll use to "create" or "update" progress
+	const tutorialProgressMutations = useTutorialProgressMutations()
+	// We don't want to run multiple mutations at once
+	const [isRunningMutation, setIsRunningMutation] = useState<boolean>(false)
```

Comment:
💭 I think we can determine this from the mutation hook return value: https://tanstack.com/query/v4/docs/reference/useMutation

In particular, `status`:

> Will be:
> `idle` initial status prior to the mutation function executing.
> `loading` if the mutation is currently executing.
> `error` if the last mutation attempt resulted in an error.
> `success` if the last mutation attempt was successful.

Reply:
@BRKalow  I've updated to include an `isReady` boolean as part of the return object from `useTutorialProgressMutations()`. Have removed `isRunningMutation` state. Let me know if that roughly lines up with what you were expecting 🙇